src/runtime/opencl/opencl_wrapper/opencl_wrapper.cc

```diff
@@ -0,0 +1,551 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file opencl_wrapper.cc
+ */
+
+#define CL_TARGET_OPENCL_VERSION 120
+#include <CL/cl.h>
+#include <CL/cl_gl.h>
+
+#define DMLC_USE_LOGGING_LIBRARY <tvm/runtime/logging.h>
+#include <dlfcn.h>
+#include <tvm/runtime/logging.h>
+
+#include <vector>
+
+namespace {
+#if defined(__APPLE__) || defined(__MACOSX)
+static const std::vector<const char*> default_so_paths = {
+    "libOpenCL.so", "/System/Library/Frameworks/OpenCL.framework/OpenCL"};
+#elif defined(__ANDROID__)
+static const std::vector<const char*> default_so_paths = {
+    "libOpenCL.so",
+    "/system/lib64/libOpenCL.so",
+    "/system/vendor/lib64/libOpenCL.so",
+    "/system/vendor/lib64/egl/libGLES_mali.so",
+    "/system/vendor/lib64/libPVROCL.so",
+    "/data/data/org.pocl.libs/files/lib64/libpocl.so",
+    "/system/lib/libOpenCL.so",
+    "/system/vendor/lib/libOpenCL.so",
+    "/system/vendor/lib/egl/libGLES_mali.so",
+    "/system/vendor/lib/libPVROCL.so",
+    "/data/data/org.pocl.libs/files/lib/libpocl.so"};
+#elif defined(_WIN32)
+static const std::vector<const char*> default_so_paths = {"OpenCL.dll"};
+#elif defined(__linux__)
+static const std::vector<const char*> default_so_paths = {"libOpenCL.so",
+                                                          "/usr/lib/libOpenCL.so",
+                                                          "/usr/local/lib/libOpenCL.so",
+                                                          "/usr/local/lib/libpocl.so",
+                                                          "/usr/lib64/libOpenCL.so",
+                                                          "/usr/lib32/libOpenCL.so"};
+#endif
+
+class LibOpenCLWrapper {
+ public:
+  static LibOpenCLWrapper& getInstance() {
+    static LibOpenCLWrapper instance;
+    return instance;
+  }
+  LibOpenCLWrapper(const LibOpenCLWrapper&) = delete;
+  LibOpenCLWrapper& operator=(const LibOpenCLWrapper&) = delete;
+  void* getOpenCLFunction(const char* funcName) {
+    if (m_libHandler == nullptr) openLibOpenCL();
+    return dlsym(m_libHandler, funcName);
+  }
+
+ private:
+  LibOpenCLWrapper() {}
+  ~LibOpenCLWrapper() {
+    if (m_libHandler) dlclose(m_libHandler);
+  }
+  void openLibOpenCL() {
+    for (const auto it : default_so_paths) {
+      m_libHandler = dlopen(it, RTLD_LAZY);
```

Comment:
this must not work on Windows. Need to branch and use LoadLibrary/GetProcAddress/FreeLibrary